proposals/interface-types/Explainer.md

```diff
@@ -0,0 +1,658 @@
+  modules.
+* **known section**: sections defined by the [core spec's binary format][Known Section]
+  which collectively decode a module
+* **custom section**: sections defined by the [core spec's binary format][Custom Section]
+  whose contents are uninterpreted by the core spec but can be interpreted by
+  other tools or specifications (including this proposal)
+
+This proposal defines the following new high-level concepts:
+* **interface type**: a new set of types that describe abstract, high-level values
+* **adapter instruction**: a new set of instructions which may produce or consume
+  values of both value types and interface types
+* **adapter function**: a new set of functions whose signatures may contain
+  interface types and whose bodies are composed of adapter instructions
+* **interface adapter**: a collection of adapter functions that are applied to a
+  module's imports and exports
+* **adapted module**: the result of applying an interface adapter to a module,
+  which encapsulates the core module's imports and exports, exposing only the
+  adapted imports and exports
+* **module interface type**: a superset of the core module type which
+  classifies an adapted module and thus includes function signatures with
+  interface types
+* **interface adapter custom section**: the new custom section defined by
+  this proposal which encodes an interface adapter which is to be applied
+  to the core module defined by the known sections to produce an adapted module
+  that shall be executed by the host in place of the core module
+
+Collectively, these concepts and their relationships can be visualized together
+in the following diagram:
+
+![Concepts overview diagram](overview.svg)
+
+
+## Walkthrough
+
+> **Note** The syntax and semantics presented below is still very much in flux.
+> Also, there's currently a large number of TODOs, so the walkthrough is
+> definitely incomplete.
+
+The walkthrough starts by introducing the `string` type and then using `string`
+in the parameters and results of first exports and then imports. Afterwards,
+other types and concepts are introduced.
+1. [Export returning string (statically allocated)](#export-returning-string-statically-allocated)
+1. [Export returning string (dynamically allocated)](#export-returning-string-dynamically-allocated)
+1. [Export receiving string](#export-receiving-string)
+1. [Lifting and lowering](#lifting-and-lowering)
+1. [Strings in imports](#strings-in-imports)
+1. [Shared-nothing linking example](#shared-nothing-linking-example)
+1. [Strings with the GC proposal](#strings-with-the-GC-proposal)
+1. [TODO](#TODO)
+
+### Export returning string (statically allocated)
+
+Let's start with a WebAssembly module that you can write today that returns a
+string that is stored at a fixed location in linear memory:
+
+```wasm
+(module
+  (memory (export "mem") 1)
+  (data (i32.const 0) "hello there")
+  (func (export "greeting_") (result i32 i32)
```

Comment:
Maybe `raw_greeting` etc.. is more clear than appending an underscore?

Reply:
Elsewhere in the proposal, I use "core" as the modifier to refer to the module that gets adapted, so I could imagine `core_greeting`; how does that sound to people?